### PR TITLE
Add visitor spec to generate visitor interface and compatible visitor method for Java

### PR DESCRIPTION
**Pull request by narenkmanoharan**
[Related issue for context](https://github.com/apollographql/apollo-android/issues/1413)

Add visitors for easier and compile-time safe consumption of polymorphic data types in the schema.

### CODE REVIEW

File: apollo-compiler/src/test/graphql/com/example/inline_fragment_for_non_optional_field/TestQuery.java [hunk 1]
```diff
@@ -177,6 +177,23 @@ public Hero read(ResponseReader reader) {
...
     ResponseFieldMarshaller marshaller();
 
+    default <T> T visit(Visitor<T> visitor) {
+      if (this instanceof AsHuman) {
+        return visitor.visit((AsHuman) this);
+      } else if (this instanceof AsCharacter) {
+        return visitor.visit((AsCharacter) this);
+      }
+      return visitor.visitDefault(this);
+    }
```
Comment (sav007) [inline apollo-compiler/src/test/graphql/com/example/inline_fragment_for_non_optional_field/TestQuery.java]: So the default implementation of function in interface will lock Apollo to be used only with Java8.
Should we guard this feature by gradle plugin configuration flag then?

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/VisitorSpec.kt [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        else -> visitorConsumer
+          .nextControlFlow(
+            "else if (this instanceof \$T)",
+            ClassName.get("", className.simpleName())
+          )
+          .addStatement(
+            "return visitor.visit((\$T) this)",
+            ClassName.get("", className.simpleName())
+          )
+      }
```
Comment (sav007) [inline apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/VisitorSpec.kt]: nit: expression 
```
.addStatement(
            "return visitor.visit((\$T) this)",
            ClassName.get("", className.simpleName())
          )
```
Can be moved out from both branches in when expression:
```
when {
}
.addStatement(
            "return visitor.visit((\$T) this)",
            ClassName.get("", className.simpleName())
          )
```

Comment (narenkmanoharan) [reply apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/VisitorSpec.kt]: done

File: apollo-compiler/src/test/graphql/com/example/inline_fragment_for_non_optional_field/TestQuery.java (hunk 1 above)
Comment (narenkmanoharan) [reply apollo-compiler/src/test/graphql/com/example/inline_fragment_for_non_optional_field/TestQuery.java]: Yeah. My thinking here was that most people on Android are already on Java 1.8 compatibility. But yeah, that's a big assumption. I'm gonna add an option in the ApolloExtension to option into generating visitors and so only those who opt into it needs to add the source/target compatibility. I'll be sure to leave a note in the readme as well.

**Pull request merged**